src/backend/screencast.rs

```diff
@@ -0,0 +1,260 @@
+    }
+}
+
+#[derive(SerializeDict, Type, Debug, Default)]
+#[zvariant(signature = "dict")]
+pub struct SelectSourcesResponse {}
+
+#[derive(DeserializeDict, Type, Debug)]
+#[zvariant(signature = "dict")]
+pub struct StartCastOptions {}
+
+#[async_trait]
+pub trait ScreencastImpl: RequestImpl + SessionImpl {
+    fn available_source_types(&self) -> BitFlags<SourceType>;
+
+    fn available_cursor_mode(&self) -> BitFlags<CursorMode>;
+
+    async fn create_session(
+        &self,
+        token: HandleToken,
+        session_token: HandleToken,
+        app_id: Option<AppID>,
+        options: CreateSessionOptions,
+    ) -> Result<CreateSessionResponse>;
+
+    async fn select_sources(
+        &self,
+        session_token: HandleToken,
+        app_id: Option<AppID>,
+        options: SelectSourcesOptions,
+    ) -> Result<SelectSourcesResponse>;
+
+    async fn start_cast(
+        &self,
+        session_token: HandleToken,
+        app_id: Option<AppID>,
+        window_identifier: Option<WindowIdentifierType>,
+        options: StartCastOptions,
+    ) -> Result<StartCastResponse>;
+}
+
+pub(crate) struct ScreencastInterface {
+    imp: Arc<dyn ScreencastImpl>,
+    cnx: zbus::Connection,
+    sessions: Arc<Mutex<SessionManager>>,
+}
+
+impl ScreencastInterface {
+    pub fn new(
+        imp: Arc<dyn ScreencastImpl>,
+        cnx: zbus::Connection,
+        sessions: Arc<Mutex<SessionManager>>,
+    ) -> Self {
+        Self { imp, cnx, sessions }
+    }
+}
+
+#[zbus::interface(name = "org.freedesktop.impl.portal.ScreenCast")]
+impl ScreencastInterface {
+    #[zbus(
+        property(emits_changed_signal = "const"),
+        name = "AvailableSourceTypes"
+    )]
+    fn available_source_types(&self) -> u32 {
+        let imp = Arc::clone(&self.imp);
+        imp.available_source_types().bits()
+    }
+
+    #[zbus(
+        property(emits_changed_signal = "const"),
+        name = "AvailableCursorModes"
+    )]
+    fn available_cursor_mode(&self) -> u32 {
+        let imp = Arc::clone(&self.imp);
+        imp.available_cursor_mode().bits()
+    }
+
+    #[zbus(property(emits_changed_signal = "const"), name = "version")]
+    fn version(&self) -> u32 {
+        5
+    }
+
+    #[zbus(name = "CreateSession")]
+    #[zbus(out_args("response", "results"))]
+    async fn create_session(
+        &self,
+        handle: OwnedObjectPath,
+        session_handle: OwnedObjectPath,
+        app_id: MaybeAppID,
+        options: CreateSessionOptions,
+    ) -> Result<Response<CreateSessionResponse>> {
+        let session_token = HandleToken::try_from(&session_handle).unwrap();
+        {
+            let sessions = self.sessions.lock().unwrap();
+            if sessions.contains(&session_token) {
+                let errormsg = format!("A session with handle `{session_token}` already exists");
+                #[cfg(feature = "tracing")]
+                tracing::error!("ScreencastInterface::create_session: {}", errormsg.as_str());
```

Comment:
```suggestion
                tracing::error!("ScreencastInterface::create_session: {}", errormsg);
```
you don't need the as_str(), String impl Display.